ocaml/typing/signature_with_global_bindings.ml

```diff
@@ -0,0 +1,74 @@
+[@@@ocaml.warning "+a-9-40-41-42"]
+
+type t = {
+  sign : Subst.Lazy.signature;
+  bound_globals : Global_module.t array;
+}
+
+let read_from_cmi (cmi : Cmi_format.cmi_infos_lazy) =
+  let sign =
+    (* Freshen identifiers bound by signature *)
+    Subst.Lazy.signature Make_local Subst.identity cmi.cmi_sign in
+  let bound_globals = cmi.cmi_globals in
+  { sign; bound_globals }
+
+let array_fold_left_filter_map f init array =
+  let ans, new_array = Array.fold_left_map f init array in
+  let new_array =
+    (* To be replaced with something faster if we need it *)
+    Array.of_seq (Seq.filter_map (fun a -> a) (Array.to_seq new_array))
+  in
+  ans, new_array
+
+let subst t (args : (Global_module.Name.t * Global_module.t) list) =
+  let { sign; bound_globals } = t in
+  match args with
+  | [] -> t
+  | _ ->
+      (* The global-level substitution *)
+      let arg_subst = Global_module.Name.Map.of_list args in
+      (* Take a bound global, substitute arguments into it, then return the
+         updated global while also adding it to the term-level substitution *)
+      let add_and_update_binding subst bound_global =
+        let name = Global_module.to_name bound_global in
+        if Global_module.Name.Map.mem name arg_subst then
+          (* This will be handled by [add_arg] below *)
```

Comment:
Well, the invariant is simply that `bound_globals` has all the globals that have hidden arguments but aren't complete. So parameters themselves currently can't show up because they can't have hidden arguments. I'll just have it error out.